### DISCUSSION TITLE
Allow use of textureSample() in compute stage

### DISCUSSION THREAD

**Discussion by jacobgorm** [gpuweb/gpuweb#4814] [Q&A] [answered]
Under HLSL and Metal I can create compute shaders that can use texture samplers to, e.g., apply an filtered image transform to a texture input coming from the camera and pass it on as a buffer to use for subsequent compute stages, e.g, for AI image analysis.
Why is the use of textureSample() restricted to fragment shaders in WGSL?

**Comment by mehmetoguzderin** [comment, marked as answer]
@jacobgorm Actually, the built-in function `textureSampleLevel` works well to sample textures in compute shaders. The `textureSample` function consumes derivatives, which are not available in compute shaders. The fact that HLSL works with `Sample` is quite recent. Until 6.6, compute shaders could only use `SampleLevel`. If you try to compile with 6.5 or lower, you get the following error today:

> error: opcode 'Derivatives in CS/MS/AS' should only be used in 'Shader Model 6.6+'.

SPIR-V lists the restriction clearly in `OpImageSampleImplicitLod`: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpImageSampleImplicitLod

However, in future, things could change. A tracking issue could be better if `textureSampleLevel` is not convenient enough for needs.

I also think that spec could have a note in the `textureSample` function to clarify that it is not available in compute shaders and refer to `textureSampleLevel` instead.
